### ISSUE TITLE
`generate_stats_figs.R` cannot find R scripts to import if used with Nextflow and non-local`work` directory

### ISSUE THREAD

**Issue by mikej888** [riboviz/riboviz#210]
If a non-standard `work` directory is used with Nextflow then the workflow fails.

```console
$ nextflow run prep_riboviz.nf -params-file local_config.yaml -work-dir /tmp/scratch/work -ansi-log false
...
[ab/4c89e7] NOTE: Process `generateStatsFigs (WTnone)` terminated with an error exit status (1) -- Error is ignored
[d7/15d618] Submitted process > generateStatsFigs (WT3AT)
[d7/15d618] NOTE: Process `generateStatsFigs (WT3AT)` terminated with an error exit status (1) -- Error is ignored
No sample was processed successfully
```

**Comment by mikej888** [comment]
`generate_stats_figs.R` fails. Check log:

```console
$ cat /tmp/scratch/work/ab/4c89e7ebe94128b79935438290a194/.command.err 
Error in file(filename, "r", encoding = encoding) : 
  cannot open the connection
Calls: source -> file
In addition: Warning message:
In file(filename, "r", encoding = encoding) :
  cannot open file '/tmp/scratch/work/ab/4c89e7ebe94128b79935438290a194/rscripts/provenance.R': No such file or directory
Execution halted
```

But `bam_to_h5.R`, which also imports `provenance.R`, succeeds;

```console
$ cat /tmp/scratch/work/76/17a0e64f31e7b443651a4212d17950/.command.out 
[1] "Created by: RiboViz"
[1] "Date: 2020-09-10 02:15:37"
[1] "File: /home/ubuntu/riboviz/rscripts/bam_to_h5.R"
[1] "Version: commit cbdffac7f0d750844d8427b54d3a70b2696437ac date 2020-08-20 15:21:24 GMT"
[1] "bam_to_h5.R running with parameters:"
$num_processes
...
```

**Comment by mikej888** [comment]
Compare source code:

* `bam_to_h5.R`:

```R
source(file.path(dirname(getopt::get_Rscript_filename()), "provenance.R"))
```

`generate_stats_figs.R`:

```R
  this_script <- getopt::get_Rscript_filename()
  path_to_this_script <- this_script
  source(here::here("rscripts", "provenance.R"))
  source(here::here("rscripts", "read_count_functions.R"))
  source(here::here("rscripts", "stats_figs_block_functions.R"))
```

The code gets the location of the current script, `generate_stats_figs.R`, but then does not use this to find the location of its peers, including `provenance.R`.

**Comment by mikej888** [comment]
See `nextflow-210` branch

**Comment by mikej888** [comment]
@FlicAnderson please could you check that this fix doesn't mess up your ability to interactively debug/run `generate_stats_figs.R`, thanks!

**Comment by FlicAnderson** [comment]
@mikej888 actually, interactive debugging/running post-nextflow is something I don't think I've cracked yet - I've previously been using the .sh command from the python run option (which I realise isn't ideal if we're moving away from that), because I run into trouble with filepaths missing for lots of file parameters... 

Nextflow strips these out, whereas with the python output for the shell script (e.g. run_riboviz_vignette.sh), all that extra info is in place.  I can just take the python created script line from that file, and change `Rscript` to `R`, change the `"/home/user/riboviz/rscripts/generate_stats_figs.R"` part to `--args`, and run that, then inside R, source the generate_stats_figs.R script. I'm not 100% sure what's the easiest way to get the equivalent from nextflow, without manually editing a number of those parameters, or trying to 'guess' the file structure above those files in the script. 

Nextflow-generated call to run generate_stats_figs.R: 
```console
Rscript --vanilla /home/fanders6/riboviz/rscripts/generate_stats_figs.R --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=WT3AT.h5 --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa --rpf=true --output-dir=. --do-pos-sp-nt-freq=true --t-rna-file=yeast_tRNAs.tsv --codon-positions-file=yeast_codon_pos_i200.RData --features-file=yeast_features.tsv --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=yeast_standard_asite_disp_length.txt --count-threshold=64
```
Compare this to the python-generated shell script output of `run_riboviz_vignette.sh` (for file parameters such as .h5 file, orf-fasta-file etc), where the extra file path information is present.: 
```console
Rscript --vanilla /home/fanders6/riboviz/rscripts/generate_stats_figs.R --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=vignette/output/WT3AT/WT3AT.h5 --orf-fasta-file=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --output-dir=vignette/output/WT3AT --do-pos-sp-nt-freq=True --t-rna-file=data/yeast_tRNAs.tsv --codon-positions-file=data/yeast_codon_pos_i200.RData --features-file=data/yeast_features.tsv --orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=data/yeast_standard_asite_disp_length.txt --count-threshold=64
```
I usually only have to edit the python version only very slightly to be able to test it interactively for this 1 sample:
```console
(riboviz) fanders6@ceres:~/riboviz$ R --vanilla --args --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=vignette/output/WT3AT/WT3AT.h5 --orf-fasta-file=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --output-dir=vignette/output/WT3AT --do-pos-sp-nt-freq=True --t-rna-file=data/yeast_tRNAs.tsv --codon-positions-file=data/yeast_codon_pos_i200.RData --features-file=data/yeast_features.tsv --orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=data/yeast_standard_asite_disp_length.txt --count-threshold=64
```
Then it runs R, with those arguments: 
```R
R version 4.0.2 (2020-06-22) -- "Taking Off Again"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
# ... various R information

> source("rscripts/generate_stats_figs.R")
[1] "Created by: RiboViz"
[1] "Date: 2020-09-14 16:43:26"
[1] "File: /home/fanders6/riboviz/rscripts/generate_stats_figs.R"
[1] "Version: commit cbdffac7f0d750844d8427b54d3a70b2696437ac date 2020-08-20 15:21:24 GMT"
[1] "generate_stats_figs.R running with parameters:"
[1] "Starting: Check for 3nt periodicity globally"
# .... the rest of the script runs fine and completes successfully and returns to the R prompt as expected. 
```

But when I try and minorly edit the nextflow version to run in R interactively: 
After this opened R, I checked what the parameters had come in as, then attempted to source generate_stats_figs.R using those details, as I have done previously, I get a failure because the path information is missing further into the script. 
```R
(riboviz) fanders6@ceres:~/riboviz$ R --vanilla --args            --num-processes=1            --min-read-length=10            --max-read-length=50            --buffer=250            --primary-id=Name            --dataset=vignette            --hd-file=WT3AT.h5            --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa            --rpf=true            --output-dir=.            --do-pos-sp-nt-freq=true            --t-rna-file=yeast_tRNAs.tsv            --codon-positions-file=yeast_codon_pos_i200.RData            --features-file=yeast_features.tsv            --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3            --asite-disp-length-file=yeast_standard_asite_disp_length.txt            --count-threshold=64

R version 4.0.2 (2020-06-22) -- "Taking Off Again"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
# ... various R information as usual

> commandArgs(trailingOnly=TRUE)
 [1] "--num-processes=1"                                            
 [2] "--min-read-length=10"                                         
 [3] "--max-read-length=50"                                         
 [4] "--buffer=250"                                                 
 [5] "--primary-id=Name"                                            
 [6] "--dataset=vignette"                                           
 [7] "--hd-file=WT3AT.h5"                                           
 [8] "--orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa"                  
 [9] "--rpf=true"                                                   
[10] "--output-dir=."                                               
[11] "--do-pos-sp-nt-freq=true"                                     
[12] "--t-rna-file=yeast_tRNAs.tsv"                                 
[13] "--codon-positions-file=yeast_codon_pos_i200.RData"            
[14] "--features-file=yeast_features.tsv"                           
[15] "--orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3"                  
[16] "--asite-disp-length-file=yeast_standard_asite_disp_length.txt"
[17] "--count-threshold=64"

# try and run generate_stats_figs.R using inputs from nextflow: 
> source("rscripts/generate_stats_figs.R")
[1] "Created by: RiboViz"
[1] "Date: 2020-09-14 15:19:10"
[1] "File: /home/fanders6/riboviz/rscripts/generate_stats_figs.R"
[1] "Version: commit 81e713986c9f0e27feffc69c0762a0d08fe7676f date 2020-09-10 09:26:46 GMT"
[1] "generate_stats_figs.R running with parameters:"
Error in h5checktypeOrOpenLoc(file, readonly = TRUE, native = native) : 
  Error in h5checktypeOrOpenLoc(). Cannot open file. File 'WT3AT.h5' does not exist.
  
# check what was the problem?  
> traceback()
7: stop("Error in ", fctname, ". Cannot open file. File '", file, 
       "' does not exist.")
6: h5checktypeOrOpenLoc(file, readonly = TRUE, native = native)
5: rhdf5::h5ls(hd_file, recursive = 1) at generate_stats_figs.R#136
4: eval(ei, envir)
3: eval(ei, envir)
2: withVisible(eval(ei, envir))
1: source("rscripts/generate_stats_figs.R")

> file.exists("vignette/output/WT3AT/WT3AT.h5")
[1] TRUE

> file.exists("WT3AT.h5")
[1] FALSE
```
Do you have any suggestions for how I/we should tackle this?  Seems that the fact that nextflow seems to intuit the "/vignette/output/WT3AT/" part of the path does mean that things won't work interactively for me without manual tweaking of paths etc...  
 
Is there anything you'd suggest for me to focus on tackling this?  Potentially I could pull the sample name off the .h5 filename, and presume that the path will be something like [dataset]/output/[sample] for the .h5 file, and the .fa and .gff files will be in something called [dataset]/input? Is there any easier or more accurate way nextflow could pass this path information accross into R perhaps? Apologies if I might be missing something obvious as my brain is still in holiday mode slightly... :palm_tree:

**Comment by mikej888** [comment]
> Nextflow strips these out, whereas with the python output for the shell script (e.g. run_riboviz_vignette.sh), all that extra info is in place.

Nextflow doesn't strip them out. Rather, they are not given to Nextflow in the first place. Nextflow runs each command in its isolated `work/` subdirectory. All the input files it needs are provided in that directory (as symlinks) and all the output files are written into that directory. Once the command is complete Nextflow copies (or "publishes") the output files to `vignette/` subdirectories but these are not read by other commands. Think of of it as Nextflow's commands as having no knowledge of `vignette/tmp` or `vignette/output` at all.

> I can just take the python created script line from that file, and change `Rscript` to `R`, change the `"/home/user/riboviz/rscripts/generate_stats_figs.R"` part to `--args`, and run that, then inside R, source the generate_stats_figs.R script. I'm not 100% sure what's the easiest way to get the equivalent from nextflow, without manually editing a number of those parameters, or trying to 'guess' the file structure above those files in the script. 

For Nextflow, each task will have a hash:

```console
$ nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false
N E X T F L O W  ~  version 20.04.1
Launching `prep_riboviz.nf` [sharp_caravaggio] - revision: 62d3b58c5d
No such sample file (NotHere): example_missing_file.fastq.gz
[35/385067] Submitted process > buildIndicesrRNA (yeast_rRNA)
...
[b6/4d6e14] Submitted process > generateStatsFigs (WTnone)
[9a/aeb496] Submitted process > generateStatsFigs (WT3AT)
...
[ce/2368b2] Submitted process > countReads
Workflow finished! (OK)
```

In this example the hash for `generateStatsFigs` on sample `WTnone` is b6/4d6e14. This provides the location of the task's directory in the Nextflow `work/` directory:

```console
$ cd work/b6/4d6e1425e1f839cbc763b6d00f8a40/
```

This includes the symbolic links to input files (denoted by `@`), output files, and logs files:

```console
$ ls -1AF
3ntframe_bygene.tsv
3ntframe_propbygene.pdf
3nt_periodicity.pdf
3nt_periodicity.tsv
codon_ribodens.pdf
codon_ribodens.tsv
.command.begin
.command.err
.command.log
.command.out
.command.run
.command.sh
.exitcode
features.pdf
pos_sp_nt_freq.tsv
pos_sp_rpf_norm_reads.pdf
pos_sp_rpf_norm_reads.tsv
read_lengths.pdf
read_lengths.tsv
startcodon_ribogridbar.pdf
startcodon_ribogrid.pdf
tpms.tsv
WTnone.h5@
yeast_codon_pos_i200.RData@
yeast_features.tsv@
yeast_standard_asite_disp_length.txt@
yeast_tRNAs.tsv@
yeast_YAL_CDS_w_250utrs.fa@
yeast_YAL_CDS_w_250utrs.gff3@
```

`.command.sh` has the exact command that Nextflow ran:

```console
$ cat .command.sh 
#!/bin/bash -ue
Rscript --vanilla /home/ubuntu/riboviz/rscripts/generate_stats_figs.R            --num-processes=1            --min-read-length=10            --max-read-length=50            --buffer=250            --primary-id=Name            --dataset=vignette            --hd-file=WTnone.h5            --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa            --rpf=true            --output-dir=.            --do-pos-sp-nt-freq=true            --t-rna-file=yeast_tRNAs.tsv            --codon-positions-file=yeast_codon_pos_i200.RData            --features-file=yeast_features.tsv            --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3            --asite-disp-length-file=yeast_standard_asite_disp_length.txt            --count-threshold=64
```

You can rerun this by doing:

```console
$ bash .command.sh
```

or by running the `Rscript` command directly:

```console
$ Rscript --vanilla /home/fanders6/riboviz/rscripts/generate_stats_figs.R ...
```

The output files are written into the current directory also.

Try interactive `R`:

```console
$ rm 3nt* codon_ribodens.* features.pdf pos_sp_* read_lengths.* startcodon_ribogrid* tpms.tsv

$ R --vanilla --args --num-processes=1            --min-read-length=10            --max-read-length=50            --buffer=250            --primary-id=Name            --dataset=vignette            --hd-file=WTnone.h5            --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa            --rpf=true            --output-dir=.            --do-pos-sp-nt-freq=true            --t-rna-file=yeast_tRNAs.tsv            --codon-positions-file=yeast_codon_pos_i200.RData            --features-file=yeast_features.tsv            --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3            --asite-disp-length-file=yeast_standard_asite_disp_length.txt            --count-threshold=64
```
```R
R version 3.6.3 (2020-02-29) -- "Holding the Windsock"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
...
Type 'q()' to quit R.

> source("../../../rscripts/generate_stats_figs.R")
```

Note, as I'm in a `work/` sub-sub-directory, I need to provide relative path up to the R script.

```R
[1] "Created by: RiboViz"
[1] "Date: 2020-09-15 02:02:56"
[1] "File: /home/ubuntu/riboviz/rscripts/generate_stats_figs.R"
[1] "Version: commit bd444db6bf84a28f8d94028e5b6d7da91cabdb4c date 2020-09-09 14:57:32 GMT"
[1] "generate_stats_figs.R running with parameters:"
[1] "Starting: Check for 3nt periodicity globally"
`summarise()` ungrouping output (override with `.groups` argument)
`summarise()` ungrouping output (override with `.groups` argument)
Saving 7 x 7 in image
[1] "Completed: Check for 3nt periodicity globally"
[1] "Starting: Distribution of lengths of all mapped reads"
Saving 7 x 7 in image
[1] "Completed: Distribution of lengths of all mapped reads"
[1] "Starting: Biases in nucleotide composition along mapped read lengths"
[1] "finished for loop"
[1] "finished prepping Length, Position, Frame variables"
[1] "finished, returning all_out"
[1] "Completed: Biases in nucleotide composition along mapped read lengths"
[1] "Starting: Check for 3nt periodicity (frame) by gene"
Parsed with column specification:
cols(
  read_length = col_double(),
  asite_displacement = col_double()
)
Note: Using an external vector in selections is ambiguous.
? Use `all_of(feat_names)` instead of `feat_names` to silence this message.
? See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.
This message is displayed once per session.
[1] "Completed: Check for 3nt periodicity (frame) by Gene"
[1] "Starting: Position specific distribution of reads"
Saving 7 x 7 in image
[1] "Completed: Position specific distribution of reads - RPF method"
[1] "Completed: Position specific distribution of reads"
[1] "Starting: Calculate TPMs of genes"
[1] "Completed: Calculate TPMs of genes"
[1] "Starting: Correlations between TPMs of genes with their sequence-based features"
Saving 7 x 7 in image
`geom_smooth()` using formula 'y ~ x'
[1] "Completed: Correlations between TPMs of genes with their sequence-based features"
[1] "Starting: Codon-specific ribosome densities for correlations with tRNAs"
Saving 7 x 7 in image
`geom_smooth()` using formula 'y ~ x'
[1] "Completed: Codon-specific ribosome densities for correlations with tRNAs"
[1] "Completed generate_stats_figs.R"
Warning messages:
1: In write.table(three_nucleotide_periodicity_data, file = tsv_file_path,  :
  appending column names to file
2: In write.table(read_length_data, file = tsv_file_path, append = T,  :
  appending column names to file
3: In write.table(all_out, file = tsv_file_path, append = T, sep = "\t",  :
  appending column names to file
4: In write.table(gene_read_frames_data, file = tsv_file_path, append = T,  :
  appending column names to file
5: In normalizePath(file_path) :
  path[1]="generate_stats_figs.R": No such file or directory
6: In write.table(pos_sp_rpf_norm_reads_data, file = tsv_file_path,  :
  appending column names to file
7: In write.table(tpms, file = tsv_file_path, append = T, sep = "\t",  :
  appending column names to file
8: Removed 8 rows containing non-finite values (stat_smooth). 
9: Removed 8 rows containing missing values (geom_point). 
10: In write.table(cod_dens_tRNA_data, file = tsv_file_path, append = T,  :
  appending column names to file
> q()
```
```console
$ ls -1AF
3ntframe_bygene.tsv
3ntframe_propbygene.pdf
3nt_periodicity.pdf
3nt_periodicity.tsv
codon_ribodens.pdf
codon_ribodens.tsv
.command.begin
.command.err
.command.log
.command.out
.command.run
.command.sh
.exitcode
features.pdf
pos_sp_nt_freq.tsv
pos_sp_rpf_norm_reads.pdf
pos_sp_rpf_norm_reads.tsv
read_lengths.pdf
read_lengths.tsv
startcodon_ribogridbar.pdf
startcodon_ribogrid.pdf
tpms.tsv
WTnone.h5@
yeast_codon_pos_i200.RData@
yeast_features.tsv@
yeast_standard_asite_disp_length.txt@
yeast_tRNAs.tsv@
yeast_YAL_CDS_w_250utrs.fa@
yeast_YAL_CDS_w_250utrs.gff3@
```

As relying on these auto-generated `work/` directories requires running Nextflow every time, one could mimic the Nextflow-style directory with symbolic links (I just thought of this today, the symbolic links fresh in my mind from riboviz/example-datasets#13!) by creating one in the `riboviz` directory:

```console
$ mkdir debug_gen_stats_figs
$ cd debug_gen_stats_figs
# Note these implicitly create symbolic links with the same names as the file names
$ ln -s ../vignette/output/WTnone/WTnone.h5
$ ln -s ../data/yeast_codon_pos_i200.RData
$ ln -s ../data/yeast_features.tsv 
$ ln -s ../data/yeast_standard_asite_disp_length.txt
$ ln -s ../data/yeast_tRNAs.tsv
$ ln -s ../vignette/input/yeast_YAL_CDS_w_250utrs.fa
$ ln -s ../vignette/input/yeast_YAL_CDS_w_250utrs.gff3
$ ls -1AF
WTnone.h5@
yeast_codon_pos_i200.RData@
yeast_features.tsv@
yeast_standard_asite_disp_length.txt@
yeast_tRNAs.tsv@
yeast_YAL_CDS_w_250utrs.fa@
yeast_YAL_CDS_w_250utrs.gff3@
```

Note: I tried doing this within the `riboviz` directory, using,

```console
$ ln -s vignette/output/WTnone/WTnone.h5 debug_gen_stats_figs/
...
```

but when trying to access the files via the symbolic links it said they didn't exist. Also, under Ubuntu the symbolic links were coloured red, but in the earlier, successful, approach they were coloured light blue.

Run `Rscript` within `debug_gen_stats_figs`:

```console
$ Rscript --vanilla /home/ubuntu/riboviz/rscripts/generate_stats_figs.R            --num-processes=1            --min-read-length=10            --max-read-length=50            --buffer=250            --primary-id=Name            --dataset=vignette            --hd-file=WTnone.h5            --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa            --rpf=true            --output-dir=.            --do-pos-sp-nt-freq=true            --t-rna-file=yeast_tRNAs.tsv            --codon-positions-file=yeast_codon_pos_i200.RData            --features-file=yeast_features.tsv            --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3            --asite-disp-length-file=yeast_standard_asite_disp_length.txt            --count-threshold=64
$ ls -1
3ntframe_bygene.tsv
3ntframe_propbygene.pdf
3nt_periodicity.pdf
3nt_periodicity.tsv
codon_ribodens.pdf
codon_ribodens.tsv
features.pdf
pos_sp_nt_freq.tsv
pos_sp_rpf_norm_reads.pdf
pos_sp_rpf_norm_reads.tsv
read_lengths.pdf
read_lengths.tsv
startcodon_ribogridbar.pdf
startcodon_ribogrid.pdf
tpms.tsv
WTnone.h5
yeast_codon_pos_i200.RData
yeast_features.tsv
yeast_standard_asite_disp_length.txt
yeast_tRNAs.tsv
yeast_YAL_CDS_w_250utrs.fa
yeast_YAL_CDS_w_250utrs.gff3
```

Try interactive `R`:

```console
$ rm 3nt* codon_ribodens.* features.pdf pos_sp_* read_lengths.* startcodon_ribogrid* tpms.tsv 
$ R --vanilla --args --num-processes=1            --min-read-length=10            --max-read-length=50            --buffer=250            --primary-id=Name            --dataset=vignette            --hd-file=WTnone.h5            --orf-fasta-file=yeast_YAL_CDS_w_250utrs.fa            --rpf=true            --output-dir=.            --do-pos-sp-nt-freq=true            --t-rna-file=yeast_tRNAs.tsv            --codon-positions-file=yeast_codon_pos_i200.RData            --features-file=yeast_features.tsv            --orf-gff-file=yeast_YAL_CDS_w_250utrs.gff3            --asite-disp-length-file=yeast_standard_asite_disp_length.txt            --count-threshold=64
```
```R
R version 3.6.3 (2020-02-29) -- "Holding the Windsock"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
...
Type 'q()' to quit R.

> source("../rscripts/generate_stats_figs.R")

[1] "Created by: RiboViz"
[1] "Date: 2020-09-15 01:56:53"
[1] "File: /home/ubuntu/riboviz/rscripts/generate_stats_figs.R"
[1] "Version: commit bd444db6bf84a28f8d94028e5b6d7da91cabdb4c date 2020-09-09 14:57:32 GMT"
[1] "generate_stats_figs.R running with parameters:"
[1] "Starting: Check for 3nt periodicity globally"
`summarise()` ungrouping output (override with `.groups` argument)
`summarise()` ungrouping output (override with `.groups` argument)
Saving 7 x 7 in image
[1] "Completed: Check for 3nt periodicity globally"
[1] "Starting: Distribution of lengths of all mapped reads"
Saving 7 x 7 in image
[1] "Completed: Distribution of lengths of all mapped reads"
[1] "Starting: Biases in nucleotide composition along mapped read lengths"
[1] "finished for loop"
[1] "finished prepping Length, Position, Frame variables"
[1] "finished, returning all_out"
[1] "Completed: Biases in nucleotide composition along mapped read lengths"
[1] "Starting: Check for 3nt periodicity (frame) by gene"
Parsed with column specification:
cols(
  read_length = col_double(),
  asite_displacement = col_double()
)
Note: Using an external vector in selections is ambiguous.
? Use `all_of(feat_names)` instead of `feat_names` to silence this message.
? See <https://tidyselect.r-lib.org/reference/faq-external-vector.html>.
This message is displayed once per session.
[1] "Completed: Check for 3nt periodicity (frame) by Gene"
[1] "Starting: Position specific distribution of reads"
Saving 7 x 7 in image
[1] "Completed: Position specific distribution of reads - RPF method"
[1] "Completed: Position specific distribution of reads"
[1] "Starting: Calculate TPMs of genes"
[1] "Completed: Calculate TPMs of genes"
[1] "Starting: Correlations between TPMs of genes with their sequence-based features"
Saving 7 x 7 in image
`geom_smooth()` using formula 'y ~ x'
[1] "Completed: Correlations between TPMs of genes with their sequence-based features"
[1] "Starting: Codon-specific ribosome densities for correlations with tRNAs"
Saving 7 x 7 in image
`geom_smooth()` using formula 'y ~ x'
[1] "Completed: Codon-specific ribosome densities for correlations with tRNAs"
[1] "Completed generate_stats_figs.R"
Warning messages:
1: In write.table(three_nucleotide_periodicity_data, file = tsv_file_path,  :
  appending column names to file
2: In write.table(read_length_data, file = tsv_file_path, append = T,  :
  appending column names to file
3: In write.table(all_out, file = tsv_file_path, append = T, sep = "\t",  :
  appending column names to file
4: In write.table(gene_read_frames_data, file = tsv_file_path, append = T,  :
  appending column names to file
5: In normalizePath(file_path) :
  path[1]="generate_stats_figs.R": No such file or directory
6: In write.table(pos_sp_rpf_norm_reads_data, file = tsv_file_path,  :
  appending column names to file
7: In write.table(tpms, file = tsv_file_path, append = T, sep = "\t",  :
  appending column names to file
8: Removed 8 rows containing non-finite values (stat_smooth). 
9: Removed 8 rows containing missing values (geom_point). 
10: In write.table(cod_dens_tRNA_data, file = tsv_file_path, append = T,  :
  appending column names to file
> q()
```
```$
$ ls -1
3ntframe_bygene.tsv
3ntframe_propbygene.pdf
3nt_periodicity.pdf
3nt_periodicity.tsv
codon_ribodens.pdf
codon_ribodens.tsv
features.pdf
pos_sp_nt_freq.tsv
pos_sp_rpf_norm_reads.pdf
pos_sp_rpf_norm_reads.tsv
read_lengths.pdf
read_lengths.tsv
startcodon_ribogridbar.pdf
startcodon_ribogrid.pdf
tpms.tsv
WTnone.h5@
yeast_codon_pos_i200.RData@
yeast_features.tsv@
yeast_standard_asite_disp_length.txt@
yeast_tRNAs.tsv@
yeast_YAL_CDS_w_250utrs.fa@
yeast_YAL_CDS_w_250utrs.gff3@
```

Hope that helps :-)

**Comment by FlicAnderson** [comment]
That's brilliant. I'll give it a go! 

I hadn't even thought about running the R debug etc from inside the nextflow work folders actually, which explains why I was missing all the context specific info and why I was failing so hard :D your explanation really helps me understand what's going on in the nextflow process better, thanks so much.

**Comment by FlicAnderson** [comment]
I've run both the options you suggested locally: running from inside the nextflow work folder, and creating the debugging folder and making the symbolic links from other places, and both work perfectly for me. I'll use these approaches for future debugging in R  - thanks! 

Since this fix branch works for me, is the next step merging/closing this issue? Or is there anything else you need me to test for it?

**Comment by ewallace** [comment]
This sounds great - which piece of documentation / developer guide do we need to update with this debugging info?

**Comment by mikej888** [comment]
I was planning to do that as a section in docs/developer/debugging.md.

**Comment by mikej888** [comment]
@FlicAnderson thanks for confirming it works!

**Comment by mikej888** [comment]
Merged into 'develop' related pull request #211.
